knitr/planetary_motion/planetary_motion.rmd

```diff
@@ -0,0 +1,735 @@
+As a prior, we use
+$$
+  k \sim \mathrm{Normal}^+(0, 1),
+$$
+which is fairly uninformative.
+
+We fit 8 chains in parallel.
+```{r }
+chains <- 8
+```
+
+```
+fit <- mod$sample(data = list(n = n, q_obs = q_obs),
+                  chains = chains, parallel_chains = chains,
+                  iter_warmup = 500,
+                  iter_sampling = 500,
+                  seed = 123, save_warmup = TRUE)
+```
+
+```{r }
+# The model takes a while to run, so we read in the saved output.
+r_fit1 <- readRDS("saved_fit/fit1.RDS")
+get_elapsed_time(r_fit1)
+```
+
+The first notable pathology is that some of the chains take a much longer time to run. The difference is not subtle...
+
+Let's examine the summary.
+```{r }
+summary(r_fit1, pars = c("lp__", "k"), probs = c())[1]
+```
+$\hat R \gg 1$. 
+Wow, these numbers are dramatic!
+Clearly the chains are not mixing and we can visualize this using trace plots.
+```{r fig.height=3}
+traceplot(r_fit1, pars = c("lp__", "k"))
+```
+
+
+Notice that chains 3 and 4, which ran for longer than 1000 seconds, sample around the largest values of $k$ and also produce the lowest log posterior distribution.
+It's not uncommon for issues to come in bulks.
+By contrast, chains that sample around $k = 1$ run in ~2 seconds, and produce a high log posterior.
+
+At this point, we may formulate several hypothesis as to what may be happening:
+
+* The model is degenerate and cannot resolve the correct value of $k$.
+* The chains get stuck at local modes and cannot escape, even if the log posterior is much higher near $k = 1$.
+* The chains get stuck on flat surfaces, and the gradient doesn't guide the chain back to $k \approx 1$.
+* The numerical integrator struggles to accurately solve the ODEs for large values of $k$. This means the ODE integrator is slower or even inaccurate. 
+In the latter case, this also means our gradient calculations, and computation of HMC could be wrong.
+
+## Is the model degenerate?
+
+Bearing a slight abuse of language, we use "degenerate" to mean that various values of $k$ roughly produce the same data generating process.
+We can check for degeneracy by looking at the  _posterior predictive checks_, split across chains.
+We plot $q_x$ against $q_y$, and for each chain, compute the median estimate for $q_\mathrm{pred}$, obtained using the `generated quantities block`.
+Note that since we fixed $\sigma = 0.01$, we expect the confidence interval to be very narrow.
+
+```{r message=FALSE}
+data_pred <- data.frame(q_obs, 1:n)
+names(data_pred) <- c("qx", "qy", "t")
+
+ppc_plot2D(r_fit1, data_pred = data_pred)
+```
+
+Clearly, only the chains that landed close to $k = 1$ (chains 2, 5, and 7) are able to fit the data.
+This is consistent with the much higher log posterior density these chains produce.
+So degeneracy alone does not explain the lack of convergence.
+Nevertheless, the chains may still be getting stuck at smaller modes, in the tail of $k$'s distribution.
```

Comment:
why?

Reply:
> This is consistent with the much higher log posterior density these chains produce. So degeneracy alone does not explain the lack of convergence. Nevertheless, the chains may still be getting stuck at smaller modes, in the tail of  
k’s distribution.

"It is not degeneracy driving the lack of convergence because otherwise all the different chains would produce roughly the same predictions."